### PR TITLE
fix: Fixed issues #69, #70 

### PR DESCRIPTION
**Pull request by Demilivor**

### CODE REVIEW

File: siocli/lib/Siocli.cpp [hunk 1]
```diff
@@ -98,9 +98,7 @@ extern "C" int siocliMain(int argc, char** argv)
         desc.add_options()("user,u",
                 boost::program_options::value<std::string>()->default_value(osUserName),
                 "User name");
-        desc.add_options()("verify-certificates,V",
-                boost::program_options::value<std::string>()->default_value(osUserName),
-                "Verify sertificates");
+        desc.add_options()("verify-certificates,V", "Verify sertificates");
```
Comment (cxxman) [inline siocli/lib/Siocli.cpp]: Verify certificates. In English, word "certificate" starts with "c", not "s".

File: siocli/lib/SqlDump.cpp [hunk 2]
```diff
@@ -42,7 +42,7 @@ struct DatabaseInfo {
     BinaryValue m_cipherKey;
 };
 
-struct ColumnConstaint {
+struct ColumnConstraint {
     ConstraintType m_type;
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp:70]: All these data structures miss doc comments for each member. REQUIRED by our DG. Please add.

File: siocli/lib/SqlDump.cpp [hunk 3]
```diff
@@ -362,7 +362,7 @@ std::vector<ColumnConstaint> receiveColumnConstraintsList(io::IoBase& connection
 
         if (rowLength == 0) break;
 
-        ColumnConstaint constraint;
+        ColumnConstraint constraint = ColumnConstraint();
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp]: Please leave this line of code as before. Instead, please introduce default constructor in the each new `struct` you've intriduced and initialize member variables of fundamental types and enums in them.

**Review by cxxman — CHANGES_REQUESTED** [review]

**Comment by cxxman** [comment]
@Demilivor Fix not accepted, please see my comments.

File: siocli/lib/SqlDump.cpp [hunk 4]
```diff
@@ -36,43 +36,111 @@ namespace {
 
 using namespace siodb::iomgr::dbengine;
 
+/**
+ * Database information
+ */
 struct DatabaseInfo {
+    /** Initializes structure DatabaseInfo. */
+    DatabaseInfo() = default;
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp]: noexcept

File: siocli/lib/SqlDump.cpp [hunk 5]
```diff
@@ -36,43 +36,111 @@ namespace {
...
+    /** Cipher key of the database */
     BinaryValue m_cipherKey;
 };
 
+/**
+ * Column constraint information
+ */
 struct ColumnConstraint {
+    /** Initializes structure ColumnConstraint. */
+    ColumnConstraint() = default;
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp]: 1. noexcept
2. Must explicitly initialize members of fundamental types and enums (in this case `m_type` and `m_constraintDefinitionId`

File: siocli/lib/SqlDump.cpp [hunk 6]
```diff
@@ -36,43 +36,111 @@ namespace {
...
+    /** Constraint expression */
     BinaryValue m_expression;
 };
 
+/**
+ * Table column information
+ */
 struct ColumnInfo {
+    /** Initializes structure ColumnInfo. */
+    ColumnInfo() = default;
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp]: 1. noexcept
2. Must explicitly initialize members of fundamental types and enums

File: siocli/lib/SqlDump.cpp [hunk 7]
```diff
@@ -36,43 +36,111 @@ namespace {
...
+    /** Constraints of this column */
     std::vector<ColumnConstraint> m_constraints;
 };
 
+/**
+ * Column set information
+ */
 struct ColumnSetInfo {
+    /** Initializes structure ColumnInfo. */
+    ColumnSetInfo() noexcept = default;
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp]: Must explicitly initialize members of fundamental types

File: siocli/lib/SqlDump.cpp [hunk 8]
```diff
@@ -36,43 +36,111 @@ namespace {
...
+    /** Column definition ID */
     std::uint64_t m_columnDefinitionId;
 };
 
+/**
+ * Column definition information
+ */
 struct ColumnDefInfo {
+    /** Initializes structure ColumnDefInfo. */
+    ColumnDefInfo() noexcept = default;
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp]: Must explicitly initialize members of fundamental types

File: siocli/lib/SqlDump.cpp [hunk 9]
```diff
@@ -36,43 +36,111 @@ namespace {
...
     std::uint64_t m_trid;
+
+    /** Column definition ID */
     std::uint64_t m_columnDefinitionId;
 };
 
+/**
+ * Column definition information
+ */
 struct ColumnDefInfo {
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp]: rename to `ColumnDefinitionInfo`

File: siocli/lib/SqlDump.cpp [hunk 10]
```diff
@@ -36,43 +36,111 @@ namespace {
...
+    /** Column ID */
     std::uint64_t m_columnId;
 };
 
+/**
+ * Table information
+ */
 struct TableInfo {
+    /** Initializes structure TableInfo. */
+    TableInfo() = default;
```
Comment (cxxman) [inline siocli/lib/SqlDump.cpp]: 1. noexcept
2. Must explicitly initialize members of fundamental types and enums

**Review by cxxman — CHANGES_REQUESTED** [review]
@Demilivor Still fix not accepted, please see more comments.

File: siocli/lib/SqlDump.cpp (hunk 4 above)
Comment (Demilivor) [reply siocli/lib/SqlDump.cpp:44]: std::string() is not noexcept

File: siocli/lib/SqlDump.cpp (hunk 5 above)
Comment (Demilivor) [reply siocli/lib/SqlDump.cpp]: BinaryValue default constructor is not noexcept

File: siocli/lib/SqlDump.cpp (hunk 6 above)
Comment (Demilivor) [reply siocli/lib/SqlDump.cpp]: BinaryValue inside ColumnConstraint  is not noexcept

**Comment by cxxman** [comment]
@Demilivor `std::string::string()` with default allocator is `noexcept` since C++ 17, because  `std::allocator::allocator()` is `noexcept`.
Look [here](https://en.cppreference.com/w/cpp/string/basic_string/basic_string) and [here](https://en.cppreference.com/w/cpp/memory/allocator/allocator).

File: siocli/lib/SqlDump.cpp (hunk 4 above)
Comment (Demilivor) [reply siocli/lib/SqlDump.cpp]: /home/alexey/siodb_demilivor/siodb/siocli/lib/SqlDump.cpp:390:22: error: use of deleted function ‘siodb::siocli::{anonymous}::DatabaseInfo::DatabaseInfo()’
         DatabaseInfo database;
                      ^~~~~~~~

BinaryValue constructor is not noexcept

**Comment by cxxman** [comment]
@Demilivor Regarding `BinaryValue`: please make following change:
1. Remove default value on size in the `stdext::buffer::buffer(std::size_t)`.
2. Add default `noexcept` constructor, initializing `m_data` and `m_end` to nullptr.

File: siocli/lib/SqlDump.cpp (hunk 4 above)
Comment (cxxman) [reply siocli/lib/SqlDump.cpp]: Please see my last 2 comments

File: siocli/lib/SqlDump.cpp (hunk 5 above)
Comment (cxxman) [reply siocli/lib/SqlDump.cpp]: Please see my last 2 comments. Make default noexcept constructor in the `stdext::buffer`

File: siocli/lib/SqlDump.cpp (hunk 6 above)
Comment (cxxman) [reply siocli/lib/SqlDump.cpp]: Please see my last 2 comments. Make default noexcept constructor in the `stdext::buffer`

**Comment by niolap** [comment]
@Demilivor 
I was able to compile on all supported OS and the 
The functionals tests were ok and I have detected those 3 points that must be fixed:

**Export <DB> should include `create database` SQL.**

`--export <DB>` must include the DDL instruction to create the database as `--export-all`.

```bash
$ siocli --user root --export db_camellia256
CREATE TABLE db_camellia256.TEST_TABLE ( CTINYINTMIN INT8, CTINYINTMAX INT8, CTINYUINT UINT8, CSMALLINTMIN INT16, CSMALLINTMAX INT16, CSMALLUINT UINT16, CINTMIN INT32, CINTMAX INT32, CUINT UINT32, CBIGINTMIN INT64, CBIGINTMAX INT64, CBIGUINT UINT64, CFLOATMIN FLOAT, CFLOATMAX FLOAT, CREALMIN DOUBLE, CREALMAX DOUBLE, CDOUBLEMIN DOUBLE, CDOUBLEMAX DOUBLE, CTEXT TEXT, CTS TIMESTAMP );

$ siocli --user root --export-all 
CREATE DATABASE DB_CAMELLIA256 WITH CIPHER_ID = 'camellia256';
CREATE TABLE DB_CAMELLIA256.TEST_TABLE ( CTINYINTMIN INT8, CTINYINTMAX INT8, CTINYUINT UINT8, CSMALLINTMIN INT16, CSMALLINTMAX INT16, CSMALLUINT UINT16, CINTMIN INT32, CINTMAX INT32, CUINT UINT32, CBIGINTMIN INT64, CBIGINTMAX INT64, CBIGUINT UINT64, CFLOATMIN FLOAT, CFLOATMAX FLOAT, CREALMIN DOUBLE, CREALMAX DOUBLE, CDOUBLEMIN DOUBLE, CDOUBLEMAX DOUBLE, CTEXT TEXT, CTS TIMESTAMP );
```

**SQL is case insensitive**

For `--export <DB>` and `--export-all` the SQL keyword and user tokens (dbs', tables' and columns' name) must be lower case (except columns values).
Currently, the DB name is either in lower or upper case and user tokens are in uppercase.

Lower cases are easier to read and SQL is case insensitive.

[Ref. here](https://stackoverflow.com/questions/292026/is-there-a-good-reason-to-use-upper-case-for-sql-keywords)

**Missing**

The header including metadata is missing from the dump but was in the initial specifications: #47 :

```sql
-- Siodb SQL dump
-- Hostname: <server_name>
-- Instance: <instance_name>
-- Timestamp: <current_timestmap e.g. 08.06.2020 12:39:12>
```

**Comment by cxxman** [comment]
@niolap Siodb internally converts all names coming from SQL clients into uppercase. Export utility reads them as is, that's why you get uppercase. Database dumps are typically not for human reading, and SQL is case insensitive, so should not matter.

**Comment by cxxman** [comment]
@niolap @Demilivor I think when exporting single database, headers must also include database name after instance:
```
-- Database: DATABASE NAME
```

**Comment by Demilivor** [comment]
@niolap 

> SQL is case insensitive
If you really want to change default case of the table name, column name, database name, 
That change should be in the siodb server. Server changes case to upper as Ivan mentioned. siocli just print the value as is.

> Export should include create database SQL.
Ok

> @niolap @Demilivor I think when exporting single database, headers must also include database name after instance:
Agree with that

> Missing
Ok, will do that

**Comment by niolap** [comment]
@Demilivor  @cxxman  Ok let's keep the current case format.

File: siocli/lib/SqlDump.cpp (hunk 4 above)
Comment (cxxman) [reply siocli/lib/SqlDump.cpp]: @Demilivor Tried to compile on centos7, and getting this error. 
So one more thing required - please change this particular constructor to 
```c++
    DatabaseInfo() noexcept
    {
    }
```
Looks like gcc-8 still have non-noexcept `std::basic_string::basic_string()`, although actually it is.
No such issue with gcc-9, so likely issue is with gcc-8 libstdc++.

File: common/lib/siodb/common/stl_ext/buffer.h [hunk 11]
```diff
@@ -43,7 +43,17 @@ class buffer {
      * Initializes object of class buffer.
      * @param size Buffer size.
      */
-    explicit buffer(size_type size = 0)
+    explicit buffer() noexcept
```
Comment (cxxman) [inline common/lib/siodb/common/stl_ext/buffer.h]: remove `explicit`

**Review by cxxman — CHANGES_REQUESTED** [review]
@Demilivor one more thing to change here

**Review by cxxman — APPROVED** [review]

**Pull request merged**